app/src/main/java/org/liberty/android/fantastischmemo/ui/QACardActivity.java

```diff
@@ -132,6 +136,54 @@ public MultipleLoaderManager getMultipleLoaderManager() {
         return multipleLoaderManager;
     }
 
+    /* favourites deck */
+    private String favouritesDbPath = "/sdcard/favourites.db";
+    private AnyMemoDBOpenHelper favouritesDbHelper;
+
+
+    public AnyMemoDBOpenHelper getFavouritesDbHelper(){
+        return favouritesDbHelper;
+    }
+
+    public void favouriteCard(Card card){
+        if(favouritesDbHelper == null) {
+            favouritesDbHelper = AnyMemoDBOpenHelperManager.getHelper(getApplicationContext(), favouritesDbPath);
+        }
+
+        favouritesDbHelper.getCardDao().createCard(card);
+    }
+
+    public void unfavouriteCard(Card card){
+        if(favouritesDbHelper == null) {
+            favouritesDbHelper = AnyMemoDBOpenHelperManager.getHelper(getApplicationContext(), favouritesDbPath);
+        }
+
+        favouritesDbHelper.getCardDao().delete(card);
+    }
+
+    public List<Card> getAllFavourites(){
+        if(favouritesDbHelper == null) {
+            favouritesDbHelper = AnyMemoDBOpenHelperManager.getHelper(getApplicationContext(), favouritesDbPath);
+        }
+
+        return favouritesDbHelper.getCardDao().queryForAll();
+    }
+
+    public void emptyFavourtiesDeck(){
+        if(favouritesDbHelper == null) {
+            favouritesDbHelper = AnyMemoDBOpenHelperManager.getHelper(getApplicationContext(), favouritesDbPath);
```

Comment:
Can you explain what exactly is happening between lines 173 and 174, where you're passing in the context the favorites db path?

Reply:
I'm getting a helper to access the favourites database. If we have already accessed the favourites deck then this helper should already exist and so we do not need to get a new one. However, if it doesn't exist yet, then we need to get it.

I had to add this to all the functions pertaining to the favourites database because there is no set order that these functions are called in. They may be called in an arbitrary order.

Reply:
Ahhhh okok makes sense!